src/posix/platform/multicast_routing.cpp

```diff
@@ -0,0 +1,564 @@
+
+        if (erase)
+        {
+            mMulticastForwardingCache.erase(it++);
+        }
+        else
+        {
+            it++;
+        }
+    }
+
+    mMulticastForwardingCache.clear();
+}
+
+void MulticastRoutingManager::ExpireMulticastForwardingCache(void)
+{
+    struct sioc_sg_req6            sioc_sg_req6;
+    std::chrono::time_point<Clock> now = Clock::now();
+    struct mf6cctl                 mf6cctl;
+
+    memset(&mf6cctl, 0, sizeof(mf6cctl));
+
+    memset(&sioc_sg_req6, 0, sizeof(sioc_sg_req6));
+
+    for (std::map<MulticastRoute, MulticastRouteInfo>::const_iterator it = mMulticastForwardingCache.begin();
+         it != mMulticastForwardingCache.end();)
+    {
+        otError                   error;
+        const MulticastRoute &    route     = it->first;
+        const MulticastRouteInfo &routeInfo = it->second;
+        bool                      erase     = false;
+
+        if ((routeInfo.mLastUseTime + std::chrono::seconds(kMulticastForwardingCacheExpireTimeout) < now))
+        {
+            if (!UpdateMulticastRouteInfo(route))
+            {
+                // The multicast route is expired
+
+                memcpy(mf6cctl.mf6cc_origin.sin6_addr.s6_addr, route.mSrcAddr.GetBytes(),
+                       sizeof(mf6cctl.mf6cc_origin.sin6_addr.s6_addr));
+                memcpy(mf6cctl.mf6cc_mcastgrp.sin6_addr.s6_addr, route.mGroupAddr.GetBytes(),
+                       sizeof(mf6cctl.mf6cc_mcastgrp.sin6_addr.s6_addr));
+
+                mf6cctl.mf6cc_parent = routeInfo.mIif;
+
+                error = (0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_DEL_MFC, &mf6cctl, sizeof(mf6cctl)))
+                            ? OT_ERROR_NONE
+                            : OT_ERROR_FAILED;
+
+                if (error == OT_ERROR_NONE || errno == ENOENT)
+                {
+                    erase = true;
+                }
+
+                otLogResultPlat(error, "MulticastRoutingManager: %s: %s => %s, MIF=%s, ForwardMIF=%s", __FUNCTION__,
+                                route.mSrcAddr.ToString().AsCString(), route.mGroupAddr.ToString().AsCString(),
+                                MifIndexToString(routeInfo.mIif), MifIndexToString(routeInfo.mOif));
+            }
+        }
+
+        if (erase)
+        {
+            mMulticastForwardingCache.erase(it++);
+        }
+        else
+        {
+            it++;
+        }
+    }
+
+    DumpMulticastForwardingCache();
+}
+
+bool MulticastRoutingManager::UpdateMulticastRouteInfo(const MulticastRoute &route)
+{
+    bool                updated = false;
+    struct sioc_sg_req6 sioc_sg_req6;
+    memset(&sioc_sg_req6, 0, sizeof(sioc_sg_req6));
+
+    memcpy(sioc_sg_req6.src.sin6_addr.s6_addr, route.mSrcAddr.GetBytes(), sizeof(sioc_sg_req6.src.sin6_addr.s6_addr));
+    memcpy(sioc_sg_req6.grp.sin6_addr.s6_addr, route.mGroupAddr.GetBytes(), sizeof(sioc_sg_req6.grp.sin6_addr.s6_addr));
+
+    MulticastRouteInfo &routeInfo = mMulticastForwardingCache[route];
```

Comment:
nit: move forward

https://github.com/openthread/openthread/blob/master/STYLE_GUIDE.md
> All scoped (i.e. stack) variable declarations should be placed together at the top of the enclosing scope in which they are used.
There shall be an empty line after all such variable declarations.